File: packages/kt-client/src/main/kotlin/dev/pyric/database/DatabaseReference.kt
package dev.pyric.database

import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import dev.pyric.database.internal.PushIdGenerator
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class DatabaseReference(
    database: FirebaseDatabase,
    path: String
) : Query(database, path) {
    private val refScope = CoroutineScope(Dispatchers.IO)

    val key: String?
        get() {
            val trimmed = path.trim('/')
            if (trimmed.isEmpty()) return null
            return trimmed.substringAfterLast('/')
        }

    val parent: DatabaseReference?
        get() {
            val trimmed = path.trim('/')
            if (trimmed.isEmpty()) return null
            val lastSlash = trimmed.lastIndexOf('/')
            val parentPath = if (lastSlash == -1) "" else trimmed.substring(0, lastSlash)
            return DatabaseReference(database, parentPath)
        }

    val root: DatabaseReference
        get() = DatabaseReference(database, "")

    fun child(pathString: String): DatabaseReference {
        val cleanChild = pathString.trim('/')
        if (cleanChild.isEmpty()) return this
        val combined = if (path.isEmpty()) cleanChild else "$path/$cleanChild"
        return DatabaseReference(database, combined)
    }

    fun push(): DatabaseReference {
        val pushKey = PushIdGenerator.generatePushId()
        return child(pushKey)
    }

    fun setValue(value: Any?): Task<Void?> = executeOp(
        method = "rtdb.set",
        params = mapOf("path" to path, "value" to value)
    )

    fun setValue(value: Any?, priority: Any?): Task<Void?> = executeOp(
        method = "rtdb.setWithPriority",
        params = mapOf("path" to path, "value" to value, "priority" to priority)
    )

    fun setPriority(priority: Any?): Task<Void?> = executeOp(
        method = "rtdb.setPriority",
        params = mapOf("path" to path, "priority" to priority)
    )

    fun updateChildren(update: Map<String, Any?>): Task<Void?> = executeOp(
        method = "rtdb.update",
        params = mapOf("path" to path, "values" to update)
    )

    fun removeValue(): Task<Void?> = executeOp(
        method = "rtdb.remove",
        params = mapOf("path" to path)
    )

    fun onDisconnect(): OnDisconnect = OnDisconnect(this)

    fun runTransaction(handler: Transaction.Handler, fireLocalEvents: Boolean = true) {
        refScope.launch {
            try {
                var maxRetries = 25
                while (maxRetries-- > 0) {
                    val currentWire = database.bridgeClient.op(
                        method = "rtdb.get",
                        params = mapOf("path" to path),
                        actAs = database.getEffectiveAuthLens().toMap()
                    )
                    @Suppress("UNCHECKED_CAST")
                    val snapMap = currentWire as? Map<String, Any?>
                    val currentVal = snapMap?.get("value")
                    val currentPriority = snapMap?.get("priority")

                    val mutableData = MutableData(currentVal, currentPriority, key)
                    val result = handler.doTransaction(mutableData)

                    if (!result.isSuccess) {
                        val currentSnap = DataSnapshot.fromWire(this@DatabaseReference, snapMap)
                        handler.onComplete(null, false, currentSnap)
                        return@launch
                    }

                    val commitRes = database.bridgeClient.op(
                        method = "rtdb.transactionCommit",
                        params = mapOf(
                            "path" to path,
                            "expected" to currentVal,
                            "value" to result.mutableData?.value,
                            "applyLocally" to fireLocalEvents
                        ),
                        actAs = database.getEffectiveAuthLens().toMap()
                    )
                    @Suppress("UNCHECKED_CAST")
                    val commitMap = commitRes as? Map<String, Any?>
                    val retry = commitMap?.get("retry") == true
                    if (!retry) {
                        @Suppress("UNCHECKED_CAST")
                        val finalSnapMap = commitMap?.get("snapshot") as? Map<String, Any?>
                        val finalSnap = DataSnapshot.fromWire(this@DatabaseReference, finalSnapMap)
                        handler.onComplete(null, commitMap?.get("committed") == true, finalSnap)
                        return@launch
                    }
                }
                handler.onComplete(
                    DatabaseError(DatabaseError.OPERATION_FAILED, "Transaction exceeded max retries"),
                    false,
                    null
                )
            } catch (e: Exception) {
                handler.onComplete(DatabaseError.fromException(e), false, null)
            }
        }
    }

    private fun executeOp(method: String, params: Map<String, Any?>): Task<Void?> {
        val tcs = TaskCompletionSource<Void?>()
        refScope.launch {
            try {
                database.bridgeClient.op(
                    method = method,
                    params = params,
                    actAs = database.getEffectiveAuthLens().toMap()
                )
                tcs.setResult(null)
            } catch (e: Exception) {
                tcs.setException(e)
            }
        }
        return tcs.task
    }

    override fun toString(): String {
        val cleanUrl = database.url.trimEnd('/')
        return if (path.isEmpty()) cleanUrl else "$cleanUrl/$path"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is DatabaseReference) return false
        return database.url == other.database.url && path == other.path
    }

    override fun hashCode(): Int {
        return 31 * database.url.hashCode() + path.hashCode()
    }
}

File: packages/ios-client/Sources/PyricDatabase/DatabaseReference.swift
import Foundation
import PyricFirestore

/// A reference to a specific location in the Realtime Database.
public final class DatabaseReference: DatabaseQuery, @unchecked Sendable {

    public init(database: Database, path: String) {
        let normalized = DatabaseReference.normalizePath(path)
        super.init(database: database, path: normalized)
    }

    internal static func normalizePath(_ raw: String) -> String {
        let parts = raw.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        if parts.isEmpty {
            return "/"
        }
        return "/" + parts.joined(separator: "/")
    }

    /// Returns the parent DatabaseReference, or nil if this is the root reference.
    public var parent: DatabaseReference? {
        if path == "/" {
            return nil
        }
        let parts = path.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        if parts.count <= 1 {
            return DatabaseReference(database: database, path: "/")
        }
        let parentPath = "/" + parts.dropLast().joined(separator: "/")
        return DatabaseReference(database: database, path: parentPath)
    }

    /// Returns the root DatabaseReference from any descendant reference.
    public var root: DatabaseReference {
        return DatabaseReference(database: database, path: "/")
    }

    /// Returns the last path segment token of the reference, or nil for the root reference.
    public var key: String? {
        if path == "/" {
            return nil
        }
        return path.split(separator: "/").last.map(String.init)
    }

    /// Returns the full URL of the reference location.
    public var url: String {
        let base = database.url.hasSuffix("/") ? String(database.url.dropLast()) : database.url
        if path == "/" {
            return base
        }
        return base + path
    }

    /// Returns a child DatabaseReference relative to the current path.
    public func child(_ pathString: String) -> DatabaseReference {
        let combined = path == "/" ? "/\(pathString)" : "\(path)/\(pathString)"
        return DatabaseReference(database: database, path: combined)
    }

    /// Generates a new child DatabaseReference with a chronologically ordered unique push ID.
    public func childByAutoId() -> DatabaseReference {
        let autoId = PushIdGenerator.generatePushId()
        return child(autoId)
    }

    // ── Write Operations ──────────────────────────────────────────────────────

    public func setValue(_ value: Any?) async throws {
        if value == nil || value is NSNull {
            try await removeValue()
            return
        }
        let params: [String: AnySendable] = [
            "path": .string(path),
            "value": RtdbValueCodec.toAnySendable(value)
        ]
        _ = try await database.bridgeClient.op(
            method: "rtdb.set",
            params: params,
            actAs: database.currentAuthLens()
        )
    }

    public func setValue(
        _ value: Any?,
        withCompletionBlock block: @escaping @Sendable (Error?, DatabaseReference) -> Void
    ) {
        let sendableValue = RtdbValueCodec.toAnySendable(value)
        Task {
            do {
                try await setValue(sendableValue)
                block(nil, self)
            } catch {
                block(error, self)
            }
        }
    }

    public func setPriority(_ priority: Any?) async throws {
        let params: [String: AnySendable] = [
            "path": .string(path),
            "priority": RtdbValueCodec.toAnySendable(priority)
        ]
        _ = try await database.bridgeClient.op(
            method: "rtdb.setPriority",
            params: params,
            actAs: database.currentAuthLens()
        )
    }

    public func setPriority(
        _ priority: Any?,
        withCompletionBlock block: @escaping @Sendable (Error?, DatabaseReference) -> Void
    ) {
        let sendablePriority = RtdbValueCodec.toAnySendable(priority)
        Task {
            do {
                try await setPriority(sendablePriority)
                block(nil, self)
            } catch {
                block(error, self)
            }
        }
    }

    public func setValue(_ value: Any?, andPriority priority: Any?) async throws {
        let params: [String: AnySendable] = [
            "path": .string(path),
            "value": RtdbValueCodec.toAnySendable(value),
            "priority": RtdbValueCodec.toAnySendable(priority)
        ]
        _ = try await database.bridgeClient.op(
            method: "rtdb.setWithPriority",
            params: params,
            actAs: database.currentAuthLens()
        )
    }

    public func setValue(
        _ value: Any?,
        andPriority priority: Any?,
        withCompletionBlock block: @escaping @Sendable (Error?, DatabaseReference) -> Void
    ) {
        let sendableValue = RtdbValueCodec.toAnySendable(value)
        let sendablePriority = RtdbValueCodec.toAnySendable(priority)
        Task {
            do {
                try await setValue(sendableValue, andPriority: sendablePriority)
                block(nil, self)
            } catch {
                block(error, self)
            }
        }
    }

    public func updateChildValues(_ values: [AnyHashable: Any]) async throws {
        let params: [String: AnySendable] = [
            "path": .string(path),
            "values": RtdbValueCodec.toAnySendable(values)
        ]
        _ = try await database.bridgeClient.op(
            method: "rtdb.update",
            params: params,
            actAs: database.currentAuthLens()
        )
    }

    public func updateChildValues(
        _ values: [AnyHashable: Any],
        withCompletionBlock block: @escaping @Sendable (Error?, DatabaseReference) -> Void
    ) {
        let sendableValues = RtdbValueCodec.toAnySendable(values)
        Task {
            do {
                let params: [String: AnySendable] = [
                    "path": .string(path),
                    "values": sendableValues
                ]
                _ = try await database.bridgeClient.op(
                    method: "rtdb.update",
                    params: params,
                    actAs: database.currentAuthLens()
                )
                block(nil, self)
            } catch {
                block(error, self)
            }
        }
    }

    public func removeValue() async throws {
        let params: [String: AnySendable] = [
            "path": .string(path)
        ]
        _ = try await database.bridgeClient.op(
            method: "rtdb.remove",
            params: params,
            actAs: database.currentAuthLens()
        )
    }

    public func removeValue(
        completionBlock block: @escaping @Sendable (Error?, DatabaseReference) -> Void
    ) {
        Task {
            do {
                try await removeValue()
                block(nil, self)
            } catch {
                block(error, self)
            }
        }
    }

    // ── Transactions ──────────────────────────────────────────────────────────

    public func runTransactionBlock(
        _ block: @escaping @Sendable (MutableData) -> TransactionResult
    ) async throws -> (committed: Bool, snapshot: DataSnapshot) {
        var attempts = 0
        let maxRetries = 25

        while attempts < maxRetries {
            attempts += 1
            let currentSnap = try await getData()
            let mutable = MutableData(
                key: currentSnap.key,
                value: currentSnap.value,
                priority: currentSnap.priority
            )

            let txResult = block(mutable)
            if txResult.isAborted {
                return (committed: false, snapshot: currentSnap)
            }

            let newValue = txResult.mutableData?.value
            let params: [String: AnySendable] = [
                "path": .string(path),
                "expected": RtdbValueCodec.toAnySendable(currentSnap.value),
                "value": RtdbValueCodec.toAnySendable(newValue),
                "applyLocally": .bool(true)
            ]

            let res = try await database.bridgeClient.op(
                method: "rtdb.transactionCommit",
                params: params,
                actAs: database.currentAuthLens()
            )

            let retry = res["retry"]?.boolValue ?? false
            if retry {
                continue
            }

            let committed = res["committed"]?.boolValue ?? true
            if let snapWire = res["snapshot"] {
                let finalSnap = DataSnapshot(wire: snapWire, ref: self)
                return (committed: committed, snapshot: finalSnap)
            } else {
                let finalSnap = try await getData()
                return (committed: committed, snapshot: finalSnap)
            }
        }

        throw PyricBridgeError.unavailable("Transaction failed after maximum retries (\(maxRetries)).")
    }

    public func runTransactionBlock(
        _ block: @escaping @Sendable (MutableData) -> TransactionResult,
        andCompletionBlock completionBlock: @escaping @Sendable (Error?, Bool, DataSnapshot?) -> Void
    ) {
        Task {
            do {
                let result = try await runTransactionBlock(block)
                completionBlock(nil, result.committed, result.snapshot)
            } catch {
                completionBlock(error, false, nil)
            }
        }
    }

    // ── OnDisconnect ──────────────────────────────────────────────────────────

    public func onDisconnect() -> OnDisconnect {
        return OnDisconnect(ref: self)
    }

    public func onDisconnectSetValue(_ value: Any?) async throws {
        try await onDisconnect().setValue(value)
    }

    public func onDisconnectSetValue(_ value: Any?, andPriority priority: Any?) async throws {
        try await onDisconnect().setValue(value, andPriority: priority)
    }

    public func onDisconnectUpdateChildValues(_ values: [AnyHashable: Any]) async throws {
        try await onDisconnect().updateChildValues(values)
    }

    public func onDisconnectRemoveValue() async throws {
        try await onDisconnect().removeValue()
    }

    public func cancelDisconnectOperations() async throws {
        try await onDisconnect().cancel()
    }
}
